File: day13_Cubicle_map_BFS/cpp/part2.cpp
#include <iostream>
#include <cstdio>
#include <vector>
#include <bitset>
#include <algorithm>
#include <map>
#include <chrono>

using namespace std;
using namespace std::chrono;

struct point{
    int x;
    int y;
};

bool checkForWall(int& x, int& y){
    int favNum = 1352;
    unsigned long productSum = x*x + 3*x + 2*x*y + y + y*y + favNum;
    bitset<16> bitSet{productSum};
    int count = bitSet.count();
    if (count % 2 == 0 && x > -1 && y > -1){
        return true;
    }else {
        return false;
    }
}

vector<point> adj(int x, int y){
    vector<vector<int>> moves {{x+1, y}, {x-1, y}, {x, y+1}, {x, y-1}};
    vector<point> allowedMoves;
    point p = {0, 0};
    for (vector<int> move : moves){
        if (checkForWall(move[0], move[1]) == true){
            p = {move[0], move[1]};
            allowedMoves.push_back(p);
        }
    }
    return allowedMoves;
}

void run(){
    point s = {1, 1};
    vector<point> frontier;
    frontier.push_back(s);

    map<int, int> level;
    level[s.x + s.y*100] = 0;

    vector<point> next;
    int i = 1;
    while(i < 51){
        next.clear();
        for (point p : frontier){
            vector<point> adjPoints = adj(p.x, p.y);
            for (point a : adjPoints){
                auto it = level.find(a.x+a.y*100);
                if (it == level.end()){
                    level[a.x+a.y*100] = i;
                    next.push_back(a);
                } else {
                }
            }
        }
        frontier = next;
        i++;
    }
    cout << level.size() << endl;
}

int main()
{
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    run();
    high_resolution_clock::time_point t2 = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>( t2 - t1 ).count();

    cout << duration << endl;
    return 0;
}

File: day13_Cubicle_map_BFS/python/part2.py
def space_check(input_number, x, y):
    z = x*x + 3*x + 2*x*y + y + y*y + input_number
    return bin(z)[2:].count('1')%2 == 0

def bfs(input_number, target):
    frontier = [[1,1]]
    locations = []
    level = 0
    while level != target and len(frontier) > 0:
        next_frontier = []
        for x,y in frontier:
            if x-1 >= 0 and space_check(input_number, x-1, y):
                if [x-1,y] not in locations:
                    next_frontier.append([x-1,y])
                    locations.append([x-1,y])
            if y-1 >= 0 and space_check(input_number, x, y-1):
                if [x,y-1] not in locations:
                    next_frontier.append([x,y-1])
                    locations.append([x,y-1])
            if space_check(input_number, x+1, y):
                if [x+1,y] not in locations:
                    next_frontier.append([x+1,y])
                    locations.append([x+1,y])
            if space_check(input_number, x, y+1):
                if [x,y+1] not in locations:
                    next_frontier.append([x,y+1])
                    locations.append([x,y+1])
        frontier = next_frontier
        level += 1
    return len(locations)

def run():
    return bfs(1352, 50)

if __name__ == "__main__":
    print(run())
